rechunker/types.py

```diff
@@ -0,0 +1,47 @@
+"""Types definitions used by executors."""
+from typing import Any, Generic, Iterable, NamedTuple, Tuple, TypeVar
+
+
+ReadableArray = Any
+WriteableArray = Any
+
+
+class CopySpec(NamedTuple):
+    """Specifcation of how to copy between two arrays."""
+
+    source: ReadableArray
+    target: WriteableArray
+    chunks: Tuple[int, ...]
+
+
+class StagedCopySpec:
+    """Specification of a copying process involving intermediate arrays.
+
+    The stages in a staged copy process must be completed in order. The
+    ``target`` of each stage can be assumed to correspond to the ``source`` of
+    the following stage.
+    """
+
+    stages: Tuple[CopySpec, ...]
+
+    def __init__(self, stages: Iterable[CopySpec]):
+        self.stages = tuple(stages)
+
+
+T = TypeVar("T")
+
+
+class Executor(Generic[T]):
```

Comment:
Can you help me understand why you didn't use an ABC here?

Reply:
I don't love ABCs because they add slow/picky runtime checks, though I guess that doesn't really matter here. Python's typing module has zero overhead, and if you don't care about type checking you don't even have to make your code pass the type checker.